com.zsmartsystems.zigbee.dongle.ember/src/main/java/com/zsmartsystems/zigbee/dongle/ember/EmberNcp.java

```diff
@@ -175,31 +173,17 @@ public class EmberNcp {
     private final Logger logger = LoggerFactory.getLogger(EmberNcp.class);
 
     /**
-     * The protocol handler used to send and receive EZSP packets
+     * The ZigBee dongle used to send and receive EZSP packets
      */
-    private final EzspProtocolHandler protocolHandler;
-
-    /**
-     * The status value from the last request
-     */
-    private EmberStatus lastStatus;
+    private final ZigBeeDongleEzsp zigBeeDongleEzsp;
 
     /**
      * Create the NCP instance
      *
-     * @param protocolHandler the {@link EzspFrameHandler} used for communicating with the NCP
-     */
-    public EmberNcp(EzspProtocolHandler protocolHandler) {
```

Comment:
Why is this removed and replaced with a call to get the protocol handler in each method?

Reply:
The protocol handler doesn't include the EZSP version required by `getVersion()`. Therefore, I replaced the `protocolHandler` property by the `zigBeeDongleEzsp` property. The `protocolHandler` can then be accessed via the `zigBeeDongleEzsp`.

Reply:
Why not save the `protocolHandler` in the constructor as well. That would allow retaining the same code in all the other methods, and reduce a set of calls to get the protocol handler every time.

Reply:
This was my first version. Then, I got an issue where the protocol handler wasn't initialized at the time EmberNcp was created. Therefore, I thought it's better not to rely on that, and access the protocol handler as late as possible (i.e. in the different methods).

Reply:
> Then, I got an issue where the protocol handler wasn't initialized at the time EmberNcp was created.

I've not looked at when your changes now create `EmberNcp`, but I guess it's created very early now as you only have a single instance that's reused? Does this issue go away if we revert to creating new instances of `EmberNcp` as per the discussion above? Or alternatively, is this only an issue due to where `EmberNcp` is instantiated, and if that was moved does that solve this issue?

Reply:
It is created only once during the initialization (`initialize()` method). The same instance is then reused all over the code.

Creating the `EmberNcp` requires some other stuffs to be initialized (like the protocol handler), therefore I found it more correct to leave this responsibility to the initialization code.

In my opinion, making the EmberNcp stateful (like before this PR) is a bad idea. What is the meaning of the last status? To which last call it refers? As I said, I still think this should be the responsibility of client code to manage this state, not the library.

Reply:
> What is the meaning of the last status? To which last call it refers?

It is the return value from the last call made to this NCP instance. Given that an NCP instance is created for each "user", this information is specific to a thread/class/method/whatever in which it was created. Therefore it has meaning to that process.